[1단계 - 자동차 경주 구현] 짱수(장혁수) 미션 제출합니다.

zangsu opened the pull request.

zangsu committed:
docs : 기능 명세 작성

zangsu committed:
feat: 이ë  횟수 입력

zangsu committed:
feat : 자동차 이름 입ì 입력 ë

zangsu committed:
feat : ì 랜덤 값을 이용해 전진 여부 결정

zangsu committed:
feat : 각 자자동차 별로 현재 위치 출력

zangsu committed:
feat : 자동차 객체 생성

zangsu committed:
feat : 위치의 최댓값 탐색

zangsu committed:
feat : 우승자 결정

zangsu committed:
feat : 최ì종 우승자 출력

zangsu committed:
feat : main 메서드 구현

zangsu committed:
refactor : 출력 메시지 추가

- 한 라운드 전체 출력
- 자동차 이름 입력 가이드 메시지 출력

zangsu committed:
feat : 이름 제약 사항 설정

zangsu committed:
refactor : RacingGame 에서 재입력을 받도록 변경

- 잘못된 자동차 이름 입력 시 재입력 요청
- 시도 횟수 재입력 위치 변경

zangsu committed:
test : Car 클래스

zangsu committed:
test : Cars 클래스

zangsu committed:
refactor : NumberGenerator 인터페이스 사용

zangsu committed:
test : Car.tryMove() 테스트

zangsu committed:
refactor : 이름 중복 방지

zangsu committed:
refactor : 컨벤션 적용

- 접근제어자 변경
- `final` 추가
- indent 1단계 고정
- 상수 추출
- 메서드 순서 변경

zangsu committed:
refactor : CarName 클래스 생성

zangsu committed:
test : Cars.getWinnersName 테스트

syoun602 commented on README.md:
```diff
@@ -2,6 +2,24 @@
+2. 이동 횟수 입력
+   1. 이동 횟수는 자연수여야 한다.
+3. 각 라운드마다 현재 상태를 출력
+   1. 랜덤 값을 이용해 전진 여부 결정
```
0~9 사이의 랜덤 값이 들어온다는 요구사항도 있네요!
기능 명세서를 조금 더 자세하게 작성해보는 것은 어떨까요?

syoun602 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,27 @@
+public class Car {
+    private static final int MOVE_THRESHOLD = 4;
+    private final CarName name;
```
매직넘버 제거 👍
상수와 인스턴스 변수 사이에 공백을 두면 한 눈에 분리해서 볼 수 있어 가독성이 더 좋아질 것 같아요 😊

syoun602 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,27 @@
+    private int location;
+    private final NumberGenerator numberGenerator;
+
+    public Car(String name, NumberGenerator numberGenerator) {
+        this.name = new CarName(name);
+        this.numberGenerator = numberGenerator;
+        location = 0;
+    }
```
`Car`이라는 객체가 `NumberGenerator`를 알아야 할까요?
해당 클래스 외부에서 자동차를 만들려고 하는데 이름 외에 `NumberGenerator`가 있으면 당황스럽지 않을까요? 😢

syoun602 commented on src/main/java/CarName.java:
```diff
@@ -0,0 +1,15 @@
+public class CarName {
```
오.. 자동차 이름에 대한 Wrapper 클래스를 두셨네요 💯

syoun602 commented on src/main/java/CarName.java:
```diff
@@ -0,0 +1,15 @@
+    public CarName(String name) {
+        if (name == null || name.isBlank() || name.length() > MAX_NAME_LENGTH) {
+            throw new IllegalArgumentException("잘못된 이름입니다.");
+        }
```
메서드로 분리해볼까요?
또한, 어떤 경우에도 같은 예외 메시지가 나타나게 되는데, 이렇게 되면 자동차 게임을 실행하는 사람 입장에서 너무 불편할 것 같아요.

syoun602 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.LinkedList;
+import java.util.List;
+
+public class Cars {
```
일급 컬렉션 활용 💯

syoun602 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,49 @@
+        if (names.isEmpty()) {
+            throw new IllegalArgumentException("이름이 입력되지 않았습니다.");
+        }
+        if (names.stream().distinct().count() != names.size()) {
```
한줄에 `stream().distinct().count()` 가 있으니 가독성이 떨어집니다!

syoun602 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,49 @@
+                .toList();
+    }
+
+    public Cars(List<Car> cars) {
+        this.cars = cars;
+    }
```
해당 메서드는 테스트에서만 사용되고 있네요.
테스트 코드 때문에 운영 코드에 추가/변경이 생기게 된다면 관리하기 힘들어질 것 같아요 😢

syoun602 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public List<String> getWinnersName() {
+        List<String> winnersName = new ArrayList<>();
```
winnerNames를 만들어 추가하는 것보다 스트림을 돌면서 한번에 list names를 꺼내볼 수 있지 않을까요?~

syoun602 commented on src/main/java/InputView.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.Scanner;
+import java.util.regex.Pattern;
+
+public class InputView {
```
정적 함수로 이루어진 클래스라면 private 생성자를 통해 다른 곳에서 의도치 않은 객체 생성을 못하도록 막아봅시다

syoun602 commented on src/main/java/InputView.java:
```diff
@@ -0,0 +1,37 @@
+    private static final Scanner sc = new Scanner(System.in);
+    private static final Pattern FINISH_WITH_DELIMITER_REGEX = Pattern.compile(".*,$");
+
+    public static List<String> readCarNames() {
```
공백에 대한 검증이 없어 실행 결과가 아래와 같던데 의도하신걸까요?
```
실행 결과
ab : 
 dd : -
 ed  : -
```

zangsu replied:
이름이 공백을 "포함"하는 경우는 정상적인 입력으로 생각하려 했습니다.
"장 혁수"와 같은 입력은 충분히 납득할 수 있는 입력이었어요.
다만, 이 과정에서 "앞, 뒤에 위치한 공백"에 대해 고민을 해 보지는 못했네요.
썬, 고맙습니다!! :D

syoun602 commented on src/main/java/InputView.java:
```diff
@@ -0,0 +1,37 @@
+                .toList();
+    }
+
+    private static boolean isFinishWithDelimiter(String input) {
+        return FINISH_WITH_DELIMITER_REGEX.matcher(input).matches();
+    }
```
정규식을 상수로 분리한 의도는 좋습니다 👍
다만, 이름에서 그 의미가 이해가 안가 다시 확인하게 되네요. 조금 더 직관적인 이름을 사용해주세요!

syoun602 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,27 @@
+    public String getName() {
+        return name.getName();
+    }
+}
```
[POSIX new line](https://blog.coderifleman.com/2015/04/04/text-files-end-with-a-newline/)에 대해 알아볼까요?
[인텔리제이 자동 개행 설정](https://kukim.tistory.com/100)도 가능합니다~!

syoun602 commented on src/main/java/RacingGame.java:
```diff
@@ -0,0 +1,38 @@
+public class RacingGame {
```
💯💯 전체적인 실행흐름이 한 눈에 볼 수 있어서 좋네요!

syoun602 commented on src/test/java/CarTest.java:
```diff
@@ -0,0 +1,73 @@
+        assertThat(car.getLocation()).isEqualTo(expectedLocation);
+    }
+
+    @ParameterizedTest(name = "{0}을 이름으로 사용하면 예외가 발생한다.")
+    @ValueSource(strings = {"", " ", "zangsu"})
+    void illegalNameExceptionTest(String name) {
+        assertThatThrownBy(() -> new Car(name, new Random()))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
```
ParameterizedTest 활용 👍
다만, 입력 값에 대한 검증을 각각 하지 않고 있어 공백에 대한 테스트인줄 알고 보다가 zzangsu가 왜 예외를 발생시키는지 헷갈릴 것 같아요. 예외 메시지를 세분화해보는건 어떨까요?

+) 이름에 대한 테스트라면 `CarNameTest`를 만들어보는건 어떨까요?

zangsu replied:
**CarNameTest에 대해...**

이름에 대한 예외 사항들을 `CarName`에서 처리해 주고 있기 때문에 `CarNameTest`에서 테스트를 해 주는 것이 적절하다는 생각이 들었고, `CarTest`에 존재하던 이름과 관련된 테스트 내용들을 `CarNameTest`로 분리했습니다.

그런데, 이 과정에서 궁금한 점이 생겼는데요.
`Car` 역시 이름을 인자로 받아 적절한 이름일 때만 생성되어야 합니다. 물론, 제가 직접 짠 코드이기에 저는 "잘못된 이름이 입력되면 `CarName`에서 예외를 발생시켜 주고, 이 예외 덕분에 `Car`에서 역시 정상적으로 예외를 넘겨 받을 수 있다."는 사실을 알고 있어요.
하지만, 제가 머릿속으로 아는 것과 코드로 테스트코드를 작성해 문서화 해 두는 것은 다른 문제라고 생각합니다.

그렇게 생각하고 나니 `CarTest`에서 역시 잘못된 이름을 사용했을 때 올바르게 예외가 발생하는지 확인해 주는 테스트가 있어야 할까? 라는 고민이 생겼습니다.

물론, 두 방법은 트레이드 오프가 존재할 것 같아요.
`CarTest`에서 잘못된 이름에 대한 테스트를 작성해 주는 것이 코드를 더 확신할 수 있도록 해준다고 생각하지만, 이를 위해 "굳이" `CarNameTest`에 있는 테스트를 한번 더 해 주는 것이 "효율적"이진 않을 것 같다고 생각해요. 동일한 내용의 테스트를 두 곳에서 중복으로 해 주기 때문에 `CarTest` 내부의 잘못된 이름에 대한 테스트는 유의미한 테스트가 되기는 어려울 것 같아요.
그럼에도, "혹시나 나중에 예외 처리 방법이 변경되었을 경우" 해당 부분의 테스트를 놓칠 가능성을 줄일 수 있지 않을까요??

정말 여러 기준들 사이에서 저만의 적절한 트레이드 오프 기준을 찾는 것은 어려운 것 같습니다. 썬의 경우 `CarTest`에서 "중복하여" 이름에 대한 예외 상황을 검증하는 것에 대해 어떻게 생각하나요?

syoun602 commented on src/test/java/CarTest.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @ParameterizedTest(name = "{0}이면 전진한다.")
+    @ValueSource(ints = {4, 5, 6, 7, 8, 9})
```
경계값 테스트 👍 요구사항에도 적어봅시다!

syoun602 commented on src/test/java/CarsTest.java:
```diff
@@ -0,0 +1,55 @@
+import org.junit.jupiter.api.Test;
+
+class CarsTest {
+    Cars cars;
+
+    @BeforeEach
+    void setup() {
+        cars = new Cars(List.of("pobi", "atto"), new Random());
+    }
```
사용되지 않고 있네요 👀

syoun602 commented on src/test/java/Constant.java:
```diff
@@ -0,0 +1,12 @@
+public class Constant implements NumberGenerator {
```
자동차 이동에 대한 테스트를 위해 잘만드셨네요 💯

`NumberGenerator`라는 이름에 맞게 변경해봅시다

syoun602 commented on src/main/java/Random.java:
```diff
@@ -0,0 +1,6 @@
+public class Random implements NumberGenerator {
```
자동차 이동 테스트를 위해 인터페이스를 분리하셨네요 💯
`NumberGenerator`라는 이름에 맞게 `Random` 클래스 명을 변경해봅시다.
현재는 자동차를 생성할 때 `Car("abc", ____ )`에 `NumberGenerator`를 넣으라고 IDE에서 추천해주는데요,
`Random`이나 `Constant`를 넣게될거라고는 생각하지 못할 것 같아요.

zangsu replied:
저도 같은 생각을 가지고 있었어요. 좀 더 정확하게는 현재의 `Random` 이라는 이름의 클래스가 `NumberGenerator` 인터페이스의 다형성을 위해 구현된 구현 클래스이기에 `NumberGenerator`를 구현했다는 걸 이름에서 더 명확히 보여 주어야 한다고 생각했어요.

그런데 제 페어의 생각은 조금 달랐어요.
어짜피 코드의 사용에서도 `NumberGenerator generator = new Random()` 과 같이 사용될 것인데, `NumberGenerator`라는 단어가 한 줄에 두 번 나타나는 것은 불필요하며, IDE의 코드 추천에서 역시 `Random` 클래스가 추천된다면 자연스럽게 해당 클래스의 역할을 이해할 수 있을 것이라고 생각했어요.

제가 페어를 납득 시킬 수 있을 만큼 클래스 이름을 짓는 기준에 대한 확신이 없었나봐요. 그래서 페어를 좀 더 설득하는 대신, 페어의 의견을 수용해 보기로 했어요. 썬의 리뷰 덕분에 해당 부분에 대해 좀 더 많이 고민해 볼 수 있었어요.

페어의 생각처럼, 어떨 때는 코드 내에서 단어가 중복해서 나타나게 될 수 있다는 말은 동의해요. 하지만 이런 상황 이외에도 정말 많은 순간에서 우리는 클래스의 역할을 클래스의 이름으로 추측하게 되는 것 같아요. 이를테면, 프로젝트를 유지/보수하는 상황에서 우리가 `Random`이라는 클래스의 이름만 보고 해당 클래스가 어떤 클래스인지 판단하기는 어려울 것 같아요.
리팩토링 역시 리뷰를 반영하는 쪽으로 진행하였습니다!

`RandomNumberGenerator` 라는 이름을 사용하는 방법이 가지는 단점에 비해, 취할 수 있는 장점이 훨씬 크다고 생각하기 때문에 저는 앞으로도 최대한 자세하게 클래스를 나타낼 수 있는 이름을 사용하게 될 것 같아요.

syoun602 replied:
좋습니다! 그렇게 자신만의 기준을 점차 쌓아나가는게 중요하다고 생각해요! 
제 생각을 덧붙이자면, `NumberGenerator generator = new Random()` 와 같은 경우 generator에 대한 구현체가 많아지면 많아질수록 각 클래스를 구분하기 힘들어 질 것 같아요. 또한, 이미 구현되어 있는 클래스들(ex. `java.util.Random`)과 이름이 중복될 가능성도 커지구요! 
따라서 단어의 길이보다는 명확하게 어떤 클래스의 구현체인지 클래스명만 보고도 알아볼 수 있는게 더 이점이 클 것 같아요.

syoun602 reviewed (changes requested):
반갑습니다 짱수 🙌
자동차 경주 리뷰어를 맡게된 썬이라고 합니다!
기본적으로 구현을 잘해주셔서 몇가지 코멘트 및 의문점을 남겨두었으니 확인하시고 궁금한게 있다면 언제든 코멘트나 DM을 활용해주세요! 

질문) 현재 패키지로 구분없이 하나의 폴더에 모든 자바 파일을 생성하셨는데, 의도가 있었을까요? 현재의 방식대로 만든다면 어떤 단점이 있을지 생각해보면 좋을 것 같아요 😊

+) PR 제목에 대한 컨벤션은 없지만 다른 사람들과 맞춰보는건 어떨까요? 또한, 구현하면서든 의문이나  집중적으로 봐줬으면 하는 부분이 있다면 description을 활용해봅시다!

zangsu committed:
refactor : 요구사항 추가

- 이동에 사용될 숫자의 범위를 명시

zangsu committed:
refactor : 가독성을 위한 리팩토링

- 상수와 인스턴스 필드 구분을 위한 개행 추가
- 생성자에서 검증 로직 분리
  - CarName.validateName()
- 조건문 내부의 복잡한 조건을 메서드로 분리
  - Cars.isDuplicated()
- CarTest 에서 이름에 관한 테스트를 CarNameTest 로 분리
- 사용되지 않는 픽스쳐 제거
  - CarsTest.cars
- NumberGenerate 구현체들 이름 변경

zangsu committed:
refactor : 파일 끝에 개행 문자 추가

zangsu committed:
refactor : 이름 앞 뒤의 공백 제거

zangsu committed:
refactor : Cars.getWinnersName 리턴 객체 변경

리스트를 생성 후 객체들을 추가하는 대신, `Stream`에 들어 있는 객체를 바로 리스트로 생성시켰음
-> 반환할 리스트의 생성과 변경이 구분되어 있다면 리스트의 수정 가능성이 열리게 되고, 예상치 못한 리스트의 수정이라는 사이드 이펙트를 막기 힘들어진다고 생각함

zangsu committed:
refactor : String.strip() 중복 제거

syoun602 commented on src/main/java/CarName.java:
```diff
@@ -3,13 +3,22 @@ public class CarName {
-        if (name == null || name.isBlank() || name.length() > MAX_NAME_LENGTH) {
-            throw new IllegalArgumentException("잘못된 이름입니다.");
-        }
+        name = name.strip();
```
Q) 사용자 입력과 관계없이 앞 뒤 공백을 지우는 것은 의도하신걸까요?!

syoun602 commented on src/main/java/CarName.java:
```diff
@@ -3,13 +3,22 @@ public class CarName {
+        }
+
+        if (name.length() > MAX_NAME_LENGTH) {
+            throw new IllegalArgumentException("이름이 너무 깁니다. 이름은 5자 이하로 입력해 주세요.");
```
세분화된 예외 메시지 👍

조금 더 나아가자면 현재 메시지의 경우, 만약 최소 시도 횟수에 대한 요구사항이 변경된다면 에러 메시지도 같이 변경되어야해요!
`String.format()`을 사용해 만들어두신 상수를 이용한다면 추후 변경에 용이할 수 있습니다! 😊

syoun602 commented on src/main/java/Cars.java:
```diff
@@ -9,14 +9,21 @@ public Cars(List<String> names, NumberGenerator numberGenerator) {
                 .toList();
     }
 
+    private static boolean isDuplicated(List<String> names) {
+        int distinctSize = (int) names.stream()
+                .distinct()
+                .count();
+        return distinctSize != names.size();
+    }
```
취향의 영역일 수도 있지만, 저라면 `validateDuplicatedNames()` 로 생성자를 조금 더 가독성 있게 만들어줄 것 같아요!

syoun602 commented on src/main/java/Cars.java:
```diff
@@ -28,22 +35,21 @@ public void tryMove() {
-        cars.stream()
+        return cars.stream()
                 .filter(car -> car.getLocation() == maxLocation)
-                .forEach(car -> winnersName.add(car.getName()));
-        return winnersName;
+                .map(Car::getName)
+                .toList();
```
훨씬 깔끔하네요 💯
filter 내부도 위치 값을 꺼내오기보다는 자동차에게 메시지를 보내볼까요?

syoun602 commented on src/main/java/Cars.java:
```diff
@@ -28,22 +35,21 @@ public void tryMove() {
     }
-}
\ No newline at end of file
+}
```
import 정리를 해줍시다!

IntelliJ에서의 자동 정렬 -> `macOS: Cmd + Opt + L`, `windows: Ctrl + Alt + L`
커밋 전에 한번씩 사용해보는 습관을 들여보아요!
[save시 자동 정렬](https://hianna.tistory.com/791)

syoun602 commented on src/test/java/CarNameTest.java:
```diff
@@ -0,0 +1,30 @@
+class CarNameTest {
+    @ParameterizedTest(name = "공백을 이름으로 사용하면 예외가 발생한다.")
+    @ValueSource(strings = {"", " "})
+    void illegalNameExceptionTest(String name) {
```
`CarNameTest` 추가 👍

> `CarTest`에서 "중복하여" 이름에 대한 예외 상황을 검증하는 것에 대해 어떻게 생각하나요?

저라면 `CarTest` 객체 내부에서 중복으로 이름에 대한 예외 상황을 테스트하지 않을 것 같아요 !
단일 책임 원칙에 의하면 각 객체의 책임은 명확하게 분리되어야 합니다. 
`Car` 객체가 주로 자동차의 동작과 상태를 관리하는 역할이라면, `CarName` 객체는 자동차의 이름과 같은 특정 데이터를 캡슐화하는 하나의 값 객체입니다.
따라서 `Car`와 `CarName`을 독립적으로 테스트하고, 각 객체의 책임에 맞는 테스트 코드를 작성함으로써 코드의 의도가 명확해지고, 유지보수성이 향상될 것 같아요. 😄

`혹시나 예외 처리 방법이 변경될 경우`는 `CarName`에 대한 테스트 코드에서 충분히 잡을 수 있지 않으까요? `Car`에서 생성했던 테스트가 정책에 맞지 않다면 테스트 코드가 통과되지 않을 것 같아서요! 

ps) 테스트 코드의 메서드명에서도 의도가 명확했으면 좋겠습니다! 물론 `DisplayName`을 통해 어떤 테스트인지 알 수 있지만, 코드 상으로도 통일되도록 테스트 코드의 목적을 명확히 전달하도록 작성해보는건 어떨까요?

syoun602 reviewed (approved):
안녕하세요 짱수~!
피드백 잘 반영해주셨네요! 자동차 경주 1단계 구현 고생 많으셨습니다 👏👏👏
1단계는 여기서 머지하고 2단계에서 또 뵐게요. 
간단한 코멘트 및 리뷰 남겨두었으니 2단계에서 반영해주세요!
(ps. 질문에 대한 답변은 리뷰 안에 답변해두었습니다.)
